### ISSUE TITLE
AS::Callbacks unreachable code

### ISSUE THREAD

**Issue by repinel** [rails/rails#21122]
After #17227 introduced a [default callback terminator](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/callbacks.rb#L525) to take care of halt conditions, it seems that [conditional](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/callbacks.rb#L281) and [simple](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/callbacks.rb#L283) methods will never be reached for `before`, `after` or `around` callbacks.

@claudiob Any thoughts here?

**Comment by claudiob** [comment]
@repinel If I remember correctly, the reason why that code looks complex and repetitive is for performance reasons, the last of which was probably https://github.com/rails/rails/issues/18011

It seems strange to think that those conditions are always skipped, and no test covers them…
but I can’t guarantee it without delving more into the code.

Could you provide more details? Do you have any test / example that I could look at?

Thanks for your feedback! :ribbon:

**Comment by repinel** [comment]
@claudiob The two test cases from https://gist.github.com/repinel/93ab39881825d913466d might illustrate my point. Using the `master` code, I don't see how one call run the callbacks without a terminator defined since there is a default. That goes against this [line](https://github.com/rails/rails/blob/3a7609e2bafee4b071fe35136274e6ccbae8cacd/activesupport/lib/active_support/callbacks.rb#L751).

If using terminators is the new default, there is no reason to keep the `conditional` and `simple` handlers in `AS::Callbacks`. I created #21173 to fix part of the doc. 

Thanks for taking the time to review this :smile:

**Comment by claudiob** [comment]
I ran the tests… added breakpoints… and now I see it! :eyes: 

In short, `chain_config.key?(:terminator)` can never be false, because if the user does not provide a terminator, we set a default one before reaching the code.

So… I think you are right :ribbon:

**Comment by kaspth** [comment]
Wasn't there also a config option which preserves the old way of halting to smoothe the upgrade? Then we can't remove these just yet.

**Comment by claudiob** [comment]
@kaspth wow! Your memory is so much better than mine!!! :sparkler: 

However, I think that @repinel might still be right. Even in case you set a value:

```ruby
ActiveSupport::Callbacks::CallbackChain.halt_and_display_warning_on_return_false = # true or false
```

then the effect is that the default terminator has [a different behavior](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/callbacks.rb#L601-L605):

```ruby
result = result_lambda.call if result_lambda.is_a?(Proc)
if halt_and_display_warning_on_return_false && result == false
  display_deprecation_warning_for_false_terminator
else
  terminate = false
end
```

but that does not change the fact that **a terminator** exists, which is all that is [being checked here](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/callbacks.rb#L129-L137):

```ruby
# This first condition can be true or false
if chain_config.key?(:terminator) && user_conditions.any?
  halting_and_conditional(callback_sequence, user_callback, user_conditions, halted_lambda, filter)

## This conditions is ALWAYS TRUE
elsif chain_config.key? :terminator
  halting(callback_sequence, user_callback, halted_lambda, filter)

## Therefore this is never reached
elsif user_conditions.any?
  conditional(callback_sequence, user_callback, user_conditions)

## Nor is this condition
else
  simple callback_sequence, user_callback
end
```

**Comment by kaspth** [comment]
@claudiob all your memory are belong to me :ghost:

Great, let's cut them and those `chain_config.key?(:terminator)` checks which are always true, :+1:.

**Comment by repinel** [comment]
@claudiob @kaspth Something still doesn't feel right. One can only get the Rails 4 behavior without setting a terminator if the config option is set to `false`.

**Comment by kaspth** [comment]
@repinel Hm, so it's still possible to hit the old code path? Then we should probably hold off on this to 5.1.

**Comment by repinel** [comment]
@kaspth No, it doesn't hit `conditional` or `simple`. It would call `halting_and_conditional ` or `halting`.

This should probably be addressed on 5.0 because if you upgrade your app from 4.2, with the default config value, your callbacks would not run properly if one of them returns `false`.

**Comment by kaspth** [comment]
@repinel So what do we do to mitigate callbacks returning false not running properly? Keep the code or remove it?
